elastic/kibana
# [OAS PoC]OAS endpoint and lazily pass schema

## Summary

Diverges a little bit from the [original PoC](https://github.com/elastic/kibana/pull/156357) but largely remains the same. The idea is to support lazily providing HTTP route schemas and also provide the ability to serve up OAS from an HTTP endpoint.

* expose the generate OAS function so that it can be used from an endpoint created in core's HTTP server
* create the ability to pass schema lazily via versioned router interface so that it can be constructed after setup/start
* use the new lazy schema option in alerting, see `x-pack/plugins/alerting/server/routes/update_rule.ts`
* also removes the CLI `node script/generate_oas` in favor of the new endpoint for now

Much of this can also be done for existing, non-versioned and versioned non-zod API routes if we add support for Joi.

## TODO

- [x] Needs to be tested end-to-end, currently the lazily loaded schemas will not be used during actual route validation

## How to test

1. Start ES & Kibana
2. Issue a request like: `curl -uelastic:changeme http://localhost:5601/<base-path-thingy>/api/oas` (optionally format with `jq` for legibility)
3. Should see a fairly incomplete OAS doc bc zod is not being used in any endpoints, but you should see one of the lazily added schemas:

```json
    "/api/alerting/rule/{id}": {
      "put": {
        "requestBody": {
          "content": {
            "application/json; Elastic-Api-Version=2023-10-31": {
              "schema": {
                "type": "object",
                "properties": {
                  "data": {
                    "type": "string",
                    "description": "I arrived lazily"
                  }
                },
                "required": [
                  "data"
                ],
                "additionalProperties": false
              }
            }
          }
        },
        "responses": {},
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string"
            }
          }
        ],
        "operationId": "/api/alerting/rule/{id}#0"
      }
    }
```
